src/wayland/data_device/server_dnd_grab.rs

```diff
@@ -182,147 +189,167 @@ where
             } else {
                 false
             };
             if let Some(ref surface) = self.current_focus {
-                for device in &seat_data.known_devices {
-                    if device.as_ref().same_client_as(surface.as_ref()) {
+                for device in seat_data.known_devices() {
+                    if device.id().same_client_as(&surface.id()) {
                         if validated {
-                            device.drop();
+                            device.drop(dh);
                         } else {
-                            device.leave();
+                            device.leave(dh);
                         }
                     }
                 }
             }
             if let Some(ref offer_data) = self.offer_data {
-                let mut data = offer_data.borrow_mut();
+                let mut data = offer_data.lock().unwrap();
                 if validated {
                     data.dropped = true;
                 } else {
                     data.active = false;
                 }
             }
-            let mut callback = self.callback.borrow_mut();
-            (&mut *callback)(ServerDndEvent::Dropped);
+            let mut callback = self.callback.lock().unwrap();
+            (*callback)(ServerDndEvent::Dropped);
             if !validated {
-                (&mut *callback)(ServerDndEvent::Cancelled);
+                (*callback)(ServerDndEvent::Cancelled);
             }
             // in all cases abandon the drop
             // no more buttons are pressed, release the grab
-            handle.unset_grab(serial, time);
+            handle.unset_grab(dh, serial, time);
         }
     }
 
-    fn axis(&mut self, handle: &mut PointerInnerHandle<'_>, details: AxisFrame) {
+    fn axis(
+        &mut self,
+        _data: &mut D,
+        dh: &mut DisplayHandle<'_>,
+        handle: &mut PointerInnerHandle<'_, D>,
+        details: AxisFrame,
+    ) {
         // we just forward the axis events as is
-        handle.axis(details);
+        handle.axis(dh, details);
     }
 
     fn start_data(&self) -> &PointerGrabStartData {
         &self.start_data
     }
 }
 
-struct OfferData {
+#[derive(Debug)]
+pub struct ServerDndOfferData {
     active: bool,
     dropped: bool,
     accepted: bool,
     chosen_action: DndAction,
 }
 
-fn implement_dnd_data_offer<C>(
-    offer: Main<wl_data_offer::WlDataOffer>,
-    metadata: super::SourceMetadata,
-    offer_data: Rc<RefCell<OfferData>>,
-    callback: Rc<RefCell<C>>,
-    action_choice: Rc<RefCell<dyn FnMut(DndAction, DndAction) -> DndAction + 'static>>,
-) -> wl_data_offer::WlDataOffer
-where
-    C: FnMut(ServerDndEvent) + 'static,
+pub fn handle_server_dnd<D>(
+    hanldler: &mut D,
+    offer: &WlDataOffer,
+    request: wl_data_offer::Request,
+
+    metadata: &super::SourceMetadata,
+    offer_data: &Arc<Mutex<ServerDndOfferData>>,
+    callback: &Arc<Mutex<dyn FnMut(ServerDndEvent) + Send + Sync>>,
```

Comment:
I would like this to be a trait based system as well. But it differs too much from client to client grabs, so you would have to use ObjectData to implement this specialized version.